### PR TITLE
feat(dockerfile): add Image Referencer for Dockerfile

### PR DESCRIPTION
**Pull request by gruebel**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

[//]: # "
    # PR Title
    Be aware that we use the title to create changelog automatically and therefore only allow specific prefixes
    - break:    to indicate a breaking change, this supersedes any of the types
    - feat:     to indicate new features or checks
    - fix:      to indicate a bugfix or handling of edge cases of existing checks
    - docs:     to indicate an update to our documentation
    - chore:    to indicate adjustments to workflow files or dependency updates
    - platform: to indicate a change needed for the platform
    Additionally a scope is needs to be added to the prefix, which indicates the targeted framework, in doubt choose 'general'.
    ex.
    feat(terraform): add CKV_AWS_123 to ensure that VPC Endpoint Service is configured for Manual Acceptance
"

## Description

- Added support for simple `FROM` instruction + multi-stage and multi-platform
- Improved type hints for Dockerfile runner, enabling `mypy` will come after my other PR is merged
- Added stubs for `dockerfile_parser` for better type hinting, will also do it for other 3rd party libs, which don't type hints over time

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my feature, policy, or fix is effective and works
- [x] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/dockerfile/image_referencer/provider.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                    # indicates a multi-platform build, therefore remove the platform flag and its value
+                    # ex. FROM --platform=$BUILDPLATFORM golang:alpine AS build
+                    multi_platform = name.split(" ", maxsplit=1)
+                    if len(multi_platform) != 2:
+                        logging.info(f"Invalid FROM statement {instruction['value']}")
+                        continue
+
+                    name = multi_platform[1]
+
+                images.append(
```
Comment (nimrodkor) [inline checkov/dockerfile/image_referencer/provider.py:49]: @gruebel I thought we wanted to take only the last FROM as it is the base image. The rest may bring CVEs with them, but this image will definitely bring CVEs that need fixing...

Comment (gruebel) [reply checkov/dockerfile/image_referencer/provider.py:49]: yeah, will adjust it 🙂

File: checkov/dockerfile/image_referencer/provider.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+from checkov.dockerfile.utils import DOCKERFILE_STARTLINE, DOCKERFILE_ENDLINE
+
+if TYPE_CHECKING:
+    from typing_extensions import TypeAlias
+
+_ExtractImagesCallableAlias: TypeAlias = Callable[["dict[str, Any]"], "list[str]"]
+
+
+class DockerfileProvider:
+    __slots__ = ("definitions", "supported_resource_types")
```
Comment (Eliran-Turgeman) [inline checkov/dockerfile/image_referencer/provider.py:17]: Is `supported_resource_types` used? Didn't find an initialization of it

**Review by Eliran-Turgeman — APPROVED** [review]
🔥

File: checkov/dockerfile/runner.py [hunk 3]
```diff
@@ -179,11 +200,23 @@ def build_record(
         record.set_guideline(check.guideline)
         report.add_record(record=record)
 
+    def extract_images(
+        self, graph_connector: DiGraph | None = None, definitions: dict[str, Any] | None = None
```
[the remark this answers is not in the document]
Comment (gruebel) [reply checkov/dockerfile/runner.py:204]: I should have merged my PR before yours 😄

File: checkov/dockerfile/image_referencer/provider.py (hunk 2 above)
Comment (gruebel) [reply checkov/dockerfile/image_referencer/provider.py:17]: good old copy paste left over 😄

**Pull request merged**